book/src/dev/state-db-upgrades.md

```diff
@@ -1,18 +1,64 @@
 # Zebra Cached State Database Implementation
 
+## Current Implementation
+
+### Verification Modes
+[verification]: #verification
+
+Zebra's state has two verification modes:
+- block hash checkpoints, and
+- full verification.
+
+This means that verification uses two different codepaths, and they must produce the same results.
+
+By default, Zebra uses as many checkpoints as it can, because they are more secure against rollbacks
+(and some other kinds of chain attacks). Then it uses full verification for the last few thousand
+blocks.
+
+When Zebra gets more checkpoints in each new release, it checks the previously verified cached
+state against those checkpoints. This checks that the two codepaths produce the same results.
+
 ## Upgrading the State Database
+[upgrades]: #upgrades
 
 For most state upgrades, we want to modify the database format of the existing database. If we
 change the major database version, every user needs to re-download and re-verify all the blocks,
 which can take days.
 
+### Writing Blocks to the State
+[write-block]: #write-block
+
+Blocks can be written to the database via two different code paths, and both must produce the same results:
+
+- Upgrading pre-existing zebra-state databases/cache to the latest format
```

Comment:
###### Optional Suggestion

Could we simplify this line to something like:

```suggestion
- Upgrading a pre-existing database to the latest format
```

Reply:
Sure!